wallet-new/src/Cardano/Wallet/Kernel/CoinSelection/FromGeneric.hs

```diff
@@ -408,4 +444,53 @@ estimateSize saa sta ins outs
 --         here with some (hopefully) realistic values.
 estimateCardanoFee :: TxSizeLinear -> Int -> [Word64] -> Word64
 estimateCardanoFee linearFeePolicy ins outs
-    = round (calculateTxSizeLinear linearFeePolicy (estimateSize 128 16 ins outs))
+    = round $ calculateTxSizeLinear linearFeePolicy
+            $ estimateSize dummyAddrAttrSize dummyTxAttrSize ins outs
+
+-- | Size to use for a value of type @Attributes AddrAttributes@ when estimating
+--   encoded transaction sizes. The minimum possible value is 2.
+dummyAddrAttrSize :: Byte
+dummyAddrAttrSize = 16
+
+-- | Size to use for a value of type @Attributes ()@ when estimating
+--   encoded transaction sizes. The minimum possible value is 2.
+dummyTxAttrSize :: Byte
+dummyTxAttrSize = 2
+
+-- | For a given transaction size, and sizes for @Attributes AddrAttributes@ and
+--   @Attributes ()@, compute the maximum possible number of inputs a transaction
+--   can have. The formula for this value is not linear, so we just do a binary
+--   search here. A decent first guess makes this search relatively quick.
+--
+--   We use a conservative over-estimate for the encoded transaction sizes, so the
+--   number of transaction inputs computed here is a lower bound on the true
+--   maximum.
+estimateMaxTxInputs
+  :: Byte -- ^ Size of @Attributes AddrAttributes@
+  -> Byte -- ^ Size of @Attributes ()@
+  -> Byte -- ^ Maximum size of a transaction
+  -> Word64
+estimateMaxTxInputs addrAttrSize txAttrSize maxSize = fromIntegral $
+  case compare (estSize txins0) maxSize of
+      LT -> bsearchUp   estSize txins0 maxSize step0
+      EQ -> txins0
+      GT -> bsearchDown estSize txins0 maxSize step0
+
+  where
+    txins0 = fromIntegral $ maxSize `div` 200
+    estSize txins = estimateSize addrAttrSize txAttrSize txins [Core.maxCoinVal]
+    step0 = 1 + (txins0 `div` 10)
+
+    bsearchUp f x y step =
+      if | f x < y -> bsearchUp f (x + step) y step
+         | f x == y -> x
+         | otherwise -> if step > 1
+                       then let step' = step `div` 2 in bsearchDown f (x - step') y step'
+                       else x - 1
```

Comment:
Rather than `MultiWayIf`, why not write as a regular ol guard?

```
bsearchUp f x y step
  | f x < y = bsearchUp f (x + step) y step
  | f x == y = x
  | otherwise = ...
```